File: haskell-solutions/src/Day8.hs
module Day8 (day8) where

import Util

day8 :: IO (Int, Int)
day8 = do
  inp <- parse
  let -- <|> for Maybe is precisely || for booleans, which is why we can
      -- take advantage of this here.
      visibles :: [Int] -> [Maybe Bool]
      visibles xs = map (\case True -> Just True; _ -> Nothing)
                  $ zipWith (>) xs (scanl' max minBound xs)
      one :: Int = length . catMaybes $ solve visibles (<|>) inp

      score :: [Int] -> [Int]
      score xs = [ maybe (length ys) succ (findIndex (>= y) ys) | (y : ys) <- tails xs]
      two :: Int = maximum $ solve score (*) inp

  pure (one, two)

-- | @solve method together xxs@ applies @method@ to every row and
-- column of @xxs@—rightwards and leftwards facing—separately, and zips
-- everything together with @together@.
solve :: forall a b. ([a] -> [b]) -> (b -> b -> b) -> [[a]] -> [b]
solve method together xxs = concat
            $ zipWith (zipWith together)
                      (map go xxs)                            -- horizontal
                      (transpose . map go . transpose $ xxs)  -- vertical
 where
  go :: [a] -> [b]
  go xs = zipWith together (method xs) (reverse . method . reverse $ xs)

parse :: IO [[Int]]
parse = map (map (read . (:[]))) . lines <$> readFile "../inputs/day8.txt"

File: clojure-solutions/src/clojure_solutions/day8.clj
(ns clojure-solutions.day8
  (:require [clojure.string :as str])
  (:use [clojure-solutions.util] :reload))

(defn- parse []
  (->> (slurp "../inputs/day8.txt")
       str/split-lines
       (mapv (comp vec (partial str-to-coll-base 10)))))

(defn- solve [trans together xxs]
  (letfn [(go [xs]                 ; 'xs' represents a single row/column
              (map together
                   (trans xs)                        ; facing one way
                   (reverse (trans (reverse xs)))))] ; facing the other way
    (apply concat
           (map (partial map together)
                (map go xxs)                         ; horizontal
                (transpose (map go (transpose xxs))) ; vertical
                ))))

(defn- find-index [p xs]
  (loop [ys xs, i 0]
    (when-not (empty? ys)
      (if (p (first ys)) i (recur (rest ys) (inc i))))))

(defn day8 [p]
  (let [input (parse)]
    (letfn [(visibles [xs]
              (map > xs (reductions max Long/MIN_VALUE xs)))
            (score [xs]
              (for [[y & ys] (drop-last (tails xs))
                    :let [i (find-index #(>= % y) ys)]]
                (if i (inc i) (count ys))))]
      (case p
        :one (reduce #(if (true? %2) (inc %1) %1)
                     0
                     (solve visibles #(or %1 %2) input))
        :two (apply max (solve score * input))))))
